kythe/docs/schema-overview.txt

```diff
@@ -0,0 +1,421 @@
+// Copyright 2018 The Kythe Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+Kythe Schema Overview
+=====================
+:Revision: 1.0
+:toc2:
+:toclevels: 3
+:priority: 999
+
+This document is intended as a high-level overview of the
+http://kythe.io/docs/schema/[Kythe graph schema], and explains some of the
+core features that make up the overall graph structure. It does not replace the
+http://kythe.io/docs/schema/[Kythe graph schema] doc, which should still be
+used as a canonical reference. Instead, this is intended to help orient the
+reader with what various pieces of the schema mean, and how to interpret it.
+
+== [[Terms]] Terminology Overview
+
+This document relies on the following basic terminology:
+
+[graphviz]
+------------------------------------------------------------------------
+digraph Terms {
+  graph [fontname="helvetica"]
+  node  [fontname="helvetica";shape=box,fontsize=10]
+  edge  [fontname="helvetica",fontsize=10]
+
+  sem [label="semantic node"]
+  anc [label="anchor";shape=ellipse]
+  A [shape=point]
+  B [shape=point]
+  A -> B [label="edge"]
+}
+------------------------------------------------------------------------
+
+An **anchor** denotes a region of a file. For example, in `kcd.go`, the
+`Reader` type/interface has many anchors:
+
+[source,Go]
+----
+type Reader interface {
+  Revisions(_ context.Context, filter *RevisionsFilter, f func(Revision) error) error
+  ^^^^^^^^^
+}
+----
+
+For example, there will be an anchor (indicated above by `^` marks) covering the
+identifier `Revisions` from `kcd.go`. The *span* of an anchor is the region of
+the file containing the entity of interest.
+
+A **semantic node** represents an abstract entity that may or may not be
+associated directly with a file. Semantic nodes obtain location by association
+with anchors.
+
+From the same example above, we can think of `Revisions` as the definition site
+of a semantic node (`function`) belonging to the type `Reader`. It has an
+anchor in `kcd.go` that defines where it is in the file. There is also another
+anchor in
+https://github.com/creachadair/kythe/blob/master/kythe/go/platform/kcd/memdb/memdb.go[`memdb.go`]
+where a `Revisions` function is implemented for the (concrete) `DB` type in
+that file.
+
+A **fact** is a named string value associated with a node. The
```

Comment:
Fact values are technically just `bytes` (not `string` with its implication of encoding).  This is only really important for `/kythe/code` and `/kythe/text`, but it should be taken into account.

Reply:
✔︎